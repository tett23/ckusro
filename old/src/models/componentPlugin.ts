import { isNonNullObject } from '../core/utils/types';
import { CkusroFile } from './CkusroFile';

export type Props = {
  currentFileId: string;
  files: CkusroFile;
};

export type ComponentPlugin = {
  name: string;
  plugin: (props: any) => JSX.Element;
};

export function isComponentPlugins(obj: any): obj is ComponentPlugin[] {
  if (!Array.isArray(obj)) {
    return false;
  }

  return obj.every(isComponentPlugin);
}

export function isComponentPlugin(obj: any): obj is ComponentPlugin {
  if (!isNonNullObject(obj)) {
    return false;
  }

  if (typeof obj.name !== 'string') {
    return false;
  }

  if (typeof obj.plugin !== 'function') {
    return false;
  }
  if (obj.plugin.length !== 1) {
    return false;
  }

  return true;
}
