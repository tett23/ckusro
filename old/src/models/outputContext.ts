import { join } from 'path';
import { isNonNullObject } from '../core/utils/types';
import { CkusroConfig } from './ckusroConfig';
import { LoaderContext } from './loaderContext';

export type OutputContext = {
  name: string;
  path: string;
};

export function newOutputContext(
  config: CkusroConfig,
  loaderContext: LoaderContext,
) {
  return {
    name: loaderContext.name,
    path: join(config.outputDirectory, loaderContext.name),
  };
}

export function isOutputContext(value: unknown): value is OutputContext {
  if (!isNonNullObject(value)) {
    return false;
  }

  if (typeof (value as OutputContext).name !== 'string') {
    return false;
  }

  if (typeof (value as OutputContext).path !== 'string') {
    return false;
  }

  return true;
}
