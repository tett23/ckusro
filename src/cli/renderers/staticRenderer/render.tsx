import React from 'react';
import { renderToString } from 'react-dom/server';
import { ServerStyleSheet, StyleSheetManager } from 'styled-components';
import App, { Props } from './assets/components';

export type RenderResult = {
  html: string;
  styles: string;
};

export default function render(props: Props): RenderResult {
  const sheet = new ServerStyleSheet();
  const html = renderToString(
    <StyleSheetManager sheet={sheet.instance}>
      <App {...props} />
    </StyleSheetManager>,
  );
  const styles = sheet.getStyleTags();

  return {
    html,
    styles,
  };
}
