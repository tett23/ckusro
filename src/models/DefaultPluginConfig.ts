import merge from 'lodash.merge';

export type DefaultPluginsConfig = {
  parsers: {
    enableWikiLink: boolean;
  };
  components: {
    enableWikiLink: boolean;
  };
};

export function defaultPluginsConfig(
  overrides: DeepPartial<DefaultPluginsConfig> = {},
): DefaultPluginsConfig {
  const data = {
    parsers: {
      enableWikiLink: true,
    },
    components: {
      enableWikiLink: true,
    },
  };

  return merge(data, overrides);
}
