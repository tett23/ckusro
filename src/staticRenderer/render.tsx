import React from 'react';
import ReactDOMServer from 'react-dom/server';
import App, { Props } from './assets/components';

export default function render(props: Props): string {
  return ReactDOMServer.renderToString(<App {...props} />);
}
