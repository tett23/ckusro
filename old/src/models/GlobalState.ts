import { dirname, join } from 'path';
import LoaderInfoBuilder from '../core/LoaderInfoBuilder';
import NodeFS from '../core/LoaderInfoBuilder/NodeFS';
import { FS } from '../core/types';
import { separateErrors } from '../core/utils/errors';
import { isErrors, isNonNullObject } from '../core/utils/types';
import { CkusroConfig } from './ckusroConfig';
import { FileBuffersState } from './FileBuffersState';
import { LoaderContext, newLoaderContexts } from './loaderContext';
import { LocalLoaderContextType } from './loaderContext/LocalLoaderContext';
import { isNamespace, Namespace } from './Namespace';
import { newOutputContext } from './OutputContext';
import { isPlugins, Plugins } from './plugins';

export type GlobalState = {
  readonly namespaces: Namespace[];
  readonly plugins: Plugins;
};

export function isGlobalState(value: unknown): value is GlobalState {
  if (!isNonNullObject(value)) {
    return false;
  }

  if (!Array.isArray((value as GlobalState).namespaces)) {
    return false;
  }

  if (!(value as GlobalState).namespaces.every(isNamespace)) {
    return false;
  }

  if (!isPlugins((value as GlobalState).plugins)) {
    return false;
  }

  return true;
}

export function newGlobalState(conf: CkusroConfig): GlobalState {
  const namespaces = newLoaderContexts(
    conf.targetDirectories,
    conf.loaderConfig,
  ).map(
    (loaderContext: LoaderContext): Namespace => {
      return {
        name: loaderContext.name,
        loaderContext,
        outputContext: newOutputContext(conf, loaderContext),
      };
    },
  );

  return {
    namespaces,
    plugins: conf.plugins,
  };
}

function getFs(context: LoaderContext): FS {
  switch (context.type) {
    case LocalLoaderContextType:
      return NodeFS;
    default:
      throw new Error();
  }
}

export async function reloadFiles(
  globalState: GlobalState,
): Promise<FileBuffersState | Error[]> {
  const { plugins } = globalState;
  const ps = globalState.namespaces.map((ns) =>
    // FIXME
    LoaderInfoBuilder(getFs(ns.loaderContext), ns, plugins),
  );
  const results: Array<FileBuffersState | Error> = (await Promise.all(
    ps,
  )).flat() as any[];
  const [fsbs, errors] = separateErrors(results);
  if (isErrors(errors)) {
    return errors;
  }

  return fsbs.reduce(
    (acc, fbs) => {
      acc.fileBuffers.push(...fbs.fileBuffers);
      acc.dependencyTable = { ...acc.dependencyTable, ...fbs.dependencyTable };
      acc.dependencyTable = {
        ...acc.invertedDependencyTable,
        ...fbs.invertedDependencyTable,
      };

      return acc;
    },
    {
      fileBuffers: [],
      dependencyTable: {},
      invertedDependencyTable: {},
    } as FileBuffersState,
  );
}

export function outputDirectory(globalState: GlobalState): string {
  return dirname(globalState.namespaces[0].outputContext.path);
}

export function assetsDirectory(globalState: GlobalState): string {
  return join(outputDirectory(globalState), 'assets');
}
