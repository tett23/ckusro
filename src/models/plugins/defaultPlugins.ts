import { DefaultPluginsConfig } from '../../models/pluginConfig';
import WikiLink from '../../plugins/ckusro-plugin-component-WikiLink/WikiLink';
import wikiLink from '../../plugins/ckusro-plugin-parser-WikiLink';
import { Plugins } from './index';

export default function defaultPlugins(config: DefaultPluginsConfig): Plugins {
  const ret: Plugins = {
    parsers: [],
    components: [],
  };

  if (config.parsers.enableWikiLink) {
    ret.parsers.push({
      name: wikiLink.name,
      plugin: wikiLink,
    });
  }

  if (config.components.enableWikiLink) {
    ret.components.push({
      name: WikiLink.name,
      plugin: WikiLink,
    });
  }

  return ret;
}
