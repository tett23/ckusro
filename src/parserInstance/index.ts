import React from 'react';
import rehypeReact from 'rehype-react';
import remarkBreaks from 'remark-breaks';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import unified from 'unified';
import { ComponentPlugin } from '../models/ComponentPlugin';
import { Plugins } from '../models/plugins';
import renderJSXComponent, { Options } from './renderJSXComponent';

export default function parserInstance(plugins: Plugins) {
  let parser = unified()
    .use(remarkParse, { gfm: true })
    .use(remarkBreaks);

  plugins.parsers.forEach(({ plugin }) => {
    parser = parser.use(plugin);
  });

  // @ts-ignore
  parser = parser.use(remarkRehype, null, jsxHandlers(plugins.components));
  parser = parser.use(rehypeReact, { createElement: React.createElement });

  return parser;
}

function jsxHandlers(plugins: ComponentPlugin[]) {
  const componentPlugins = plugins.reduce(
    (acc, { plugin }) => {
      acc[plugin.name || plugin.name] = plugin;

      return acc;
    },
    {} as any,
  );
  const remarkResolveJSXOptions: Options = {
    components: componentPlugins,
  };
  const handlers = {
    jsx: (_: any, node: any) => {
      return renderJSXComponent(remarkResolveJSXOptions, node);
    },
  };

  return {
    handlers,
  };
}
