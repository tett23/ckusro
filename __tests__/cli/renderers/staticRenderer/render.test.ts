/**
 * @jest-environment node
 */

import render from '../../../../src/cli/renderers/staticRenderer/render';
import { defaultPluginsConfig } from '../../../../src/models/DefaultPluginConfig';
import defaultPlugins from '../../../../src/models/plugins/defaultPlugins';
import parserInstance from '../../../../src/parserInstance';
import { buildFileBufferState } from '../../../__fixtures__';

describe(render, () => {
  it('renders correctly', () => {
    const plugins = defaultPlugins(defaultPluginsConfig());
    const fbState = buildFileBufferState();
    const actual = render({
      common: {
        parserInstance: parserInstance(plugins),
      },
      fileBuffers: {
        fileBuffersState: fbState,
        currentFileBufferId: fbState.fileBuffers[0].id,
      },
    });

    expect(actual).toMatchSnapshot();
  });
});
