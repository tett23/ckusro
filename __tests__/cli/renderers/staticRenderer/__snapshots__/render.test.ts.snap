// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
Object {
  "html": "<main class=\\"sc-ifAKCX cisiRd\\"><nav class=\\"sc-bxivhb hYiooe\\"><ul></ul></nav><div class=\\"sc-htpNat hCKVJH\\"><nav><nav><ul class=\\"sc-bwzfXH hdWBic\\"><li class=\\"sc-bdVaJa cmHtWv\\"></li></ul></nav></nav><div><section><p><span data-reactroot=\\"\\">bar.md</span></p></section><section><pre>[[bar.md]]</pre></section></div></div></main>",
  "styles": "<style data-styled=\\"cisiRd hYiooe hCKVJH hdWBic cmHtWv\\" data-styled-version=\\"4.1.3\\">
/* sc-component-id: sc-bdVaJa */
.cmHtWv:after{padding-left:0.5rem;padding-right:0.5rem;content:'/';} .cmHtWv:last-child:after{content:'';}
/* sc-component-id: sc-bwzfXH */
.hdWBic{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;list-style:none;padding:0;margin:0;}
/* sc-component-id: sc-htpNat */
.hCKVJH{-webkit-flex:auto;-ms-flex:auto;flex:auto;}
/* sc-component-id: sc-bxivhb */
.hYiooe{-webkit-flex:1;-ms-flex:1;flex:1;}
/* sc-component-id: sc-ifAKCX */
.cisiRd{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;}</style>",
}
`;
